### ISSUE TITLE
Disagreement with "Encyclopedia of emergent ..." paper in space groups 168-173

### ISSUE THREAD

**Issue by thchr** [thchr/KdotP.jl#5]
Our check for `isweyl` does agree with the results of Table S1 in https://doi.org/10.1016/j.scib.2021.10.023 for space groups 168-173 (see the [supplement](https://ars.els-cdn.com/content/image/1-s2.0-S2095927321006927-mmc1.pdf)).

It seems that there is something off about the time-reversal constraint in our implementation, since all these cases are cases where we have irreps that are glued together.

E.g., for space group 168, we currently get:
```jl
julia> lgirsd = lgirreps(168)
julia> lgirsd_tr = Dict(klab=>realify(lgirs) for (klab, lgirs) in lgirsd)
julia> kdotp(lgirsd_tr["K"][2]; timereversal=true) # the K2K3 irrep
HamiltonianExpansion{3} up to degree 1 for 2D irrep (K₂K₃):
┌ MonomialHamiltonian{3} of degree 1 with 1 basis elements:
│ ₁₎ ┌       ┐ 
│    │ 1   · │z
│    │ ·  -1 │ 
└    └       ┘ 
```
Interestingly, if we pretend the irrep is glued together but not a corep - i.e., ignore the time-reversal constraint - we get a more Weyl-like result:
```jl
julia> kdotp(lgirsd["K"][2]+lgirsd["K"][3]; timereversal=false)
HamiltonianExpansion{3} up to degree 1 for 2D irrep (K₂⊕K₃):
┌ MonomialHamiltonian{3} of degree 1 with 4 basis elements:
│ ₁₎ ┌      ┐          ┌       ┐         
│    │ ·  1 │(-2x+y) + │ ·  -i │(-1.732y)
│    │ 1  · │          │ i   · │         
│    └      ┘          └       ┘         
│ ₂₎ ┌      ┐         ┌       ┐       
│    │ ·  1 │1.732y + │ ·  -i │(-2x+y)
│    │ 1  · │         │ i   · │       
│    └      ┘         └       ┘       
│ ₃₎ ┌       ┐ 
│    │ 1   · │z
│    │ ·  -1 │ 
│    └       ┘ 
│ ₄₎ ┌      ┐ 
│    │ 1  · │z
│    │ ·  1 │ 
└    └      ┘
```
The last term disagrees with the model listed in the supplement of https://doi.org/10.1016/j.scib.2021.10.023 (p. 696) - but otherwise it seems about right.

**Comment by thchr** [comment]
I think there could be something wrong with the antiunitary irrep we identify.

In particular, for the above example of `lgirsd_tr["K"][2]` (K2K3, SG 168), things turn out for the better if we swap the antiunitary irrep that is returned by `find_antiunitary_irrep` -> `find_unitary_transform_bc`, which returns
```jl
2×2 Matrix{ComplexF64}:
      -0.0+1.04673e-16im  0.707107+0.707107im
 -0.707107+0.707107im         -0.0-1.04673e-16i
```
for the one returned by  `find_antiunitary_irrep` -> `find_unitary_transform` (no `_bc`):
```jl
2×2 Matrix{ComplexF64}:
 -0.0+0.0im  -0.0+1.0im
 -0.0+1.0im  -0.0+0.0im
```
In principle, both seem to be acceptable choices - but when we impose the constraint via the former, things go awry (but not via the latter).

EDIT: To verify this, do the following before calculating the k-dot-p model:
```jl
@eval KdotP function find_antiunitary_irrep(lgir, g₀)
    Xs = lgir(nothing)                               # D(gᵢ)
    Ys = trs_transformed_irreps(Xs, group(lgir), g₀) # D*(g₀gᵢg₀⁻¹)

    if all(((X,Y),) -> tr(X) ≈ tr(Y), zip(Xs, Ys))
        D_𝒯g₀ = #try 
            #find_unitary_transform_bc(Xs, Ys)
        #catch 
            find_unitary_transform(Xs, Ys)
        #end
        return D_𝒯g₀
    else
        # TODO: The irrep is complex; this should probably not be possible if fed by a
        #       corep, but would be nice to generalize our scope here
        error("Xs and Ys are not equivalent: don't know how to proceed")
    end
end
```

**Comment by thchr** [comment]
The fundamental issue is that, depending on how `find_antiunitary_irrep` resolves the issue - via either `find_unitary_transform_bc` or `find_unitary_transform` (and even, with what "random-matrix" initialization for the former) - it can find different answers to what the unitary part of the corep D_𝒯g₀ should be.

In the SG168 case, the current (and wrong) situation is that it finds the following via `find_unitary_transform_bc`:
```jl
D1 = [0 cis(2π/8); cis(2π/8+π/2) 0] 
```
While `find_unitary_transform` would return:
```jl
D2 =  [0 1im; 1im 0.0]
```
`D1` is nominally okay at first glance; in particular, it _does_ fulfil the requirement we impose on it, from Eq. (S58) of Bradlyn, Science (2016). Same is true for `D2`. 
**BUT**, it seems the requirement from (S58) is not sufficient. In particular, we would also require the following, cf. its antiunitary nature:

$$
D(𝒯g_0)D^*(𝒯g_0) = D(g_0g_0)
$$

Which, in the present case where g₀ is a 2-fold rotation so that $g_0g_0 = 1$, implies $D(𝒯g_0)D^*(𝒯g_0)= D(1) = 1$.
But that is not what we get for `D1` (but `D2 == I(2)` holds):
```jl
julia> D1*conj(D1)
2×2 Matrix{ComplexF64}:
 0.0-1.0im  0.0+0.0im
 0.0+0.0im  0.0+1.0im
```
(NB: but, both `D1` and `D2` are unitary)

This suggests that it is not enough to only impose Bradlyn's (S58) in general: with only that, we might get something that does not fulfil all our requirements for the algebra of coreps [i.e., Bradley & Cracknell's Eqs. (7.3.19)].

We can take a step back and think about whether this is reasonable. In the present case, gᵢ can range over {1,  3₀₀₁⁺, 3₀₀₁⁻} and Bradlyn's Eq. (S58) requires:
> D(𝒯g₀) D(gᵢ) D(𝒯g₀ )⁻¹  =  D*(g₀gᵢg₀⁻¹)

And we also have here that that D*(g₀gᵢg₀⁻¹) = D*(gᵢ) since g₀ = 2₀₀₁.
Notably, the gᵢ = 1 equation does not give us any information since it collapses to just D(𝒯g₀) D(𝒯g₀ )⁻¹ = D(1) = 1. So, we "just" have two 2x2 by equations from S58 - it's not obvious that this gives us enough relations to pin down D(𝒯g₀) completely. In fact, it does not at all: in particular, we can choose D(𝒯g₀) as any matrix of the following form:

D(𝒯g₀) = [0 c; d 0]

Of course, then it is not necessarily unitary - nor special unitary - but a special case of the above which is in SU(2) and still has one free parameter is

D(𝒯g₀) = [0 exp(iφ); -exp(-iφ) 0]

which reduces to `D1` for φ=π/4 and to `D2` for φ=π/2. So, basically, the mystery is not that hard: we don't have enough conditions imposed on D(𝒯g₀) to pin all its degrees of freedom down - and we don't include enough constraints to make sure we satisfy the required algebra.

In particular, if we had included the constraint D(𝒯g_0)D^*(𝒯g_0) = D(g_0g_0) = 1, we would have pinned down φ = π/2+πn.

**CONCLUSION:** We need to impose more constraints in order to get the right result.

**Comment by thchr** [comment]
Fixed by [2db4943](https://github.com/thchr/KdotP.jl/commit/2db4943a041fdbec9eaaba30b90f177aa5fff841)
